NAMD parser does not take temperature as argument

The NAMD parser #7 (for FEP) does not take `T` as an argument and does not return `u_nk` in proper units of kT.

@orbeckst I'm trying to understand where the error is coming from. In the Travis message there are the lines:

```
>       assert X_delta_f[1] == pytest.approx(delta_f, rel=1e-3)
E       assert 10.116843440154604 == 11.0044412521 +- 1.1e-02
```

where 10.117 is approximately the input value for the test from 6.031 kcal/mol, so is this message saying that the internal test is returning about 11 kT?

--- a/setup.py
+++ b/setup.py
@@ -29,5 +29,5 @@ setup(name='alchemlyb',
       license='BSD',
       long_description=open('README.rst').read(),
       tests_require = ['pytest', 'alchemtest'],
-      install_requires=['numpy', 'pandas', 'pymbar', 'scipy', 'scikit-learn']
+      install_requires=['numpy', 'pandas>=0.23.0', 'pymbar', 'scipy', 'scikit-learn']
       )

--- a/src/alchemlyb/parsing/namd.py
+++ b/src/alchemlyb/parsing/namd.py
@@ -30,69 +30,67 @@ def extract_u_nk(fep_file, T):
     """
     beta = 1/(k_b * T)
 
-    # lists to get timesteps and work values of each window
+    # lists to get times and work values of each window
     win_ts = []
     win_de = []
 
     # create dataframe for results
-    u_nk = pd.DataFrame(columns=['timestep','fep-lambda'])
+    u_nk = pd.DataFrame(columns=['time','fep-lambda'])
 
     # boolean flag to parse data after equil time
     parsing = False
 
     # open and get data from fep file.
     with anyopen(fep_file, 'r') as f:
-        data = f.readlines()
+        for line in f:
+            l = line.strip().split()
 
-    for line in data:
-        l = line.strip().split()
+            # this line marks end of window; dump data into dataframe
+            if '#Free' in l:
 
-        # this line marks end of window; dump data into dataframe
-        if '#Free' in l:
+                # convert last window's work and times values to np arrays
+                win_de_arr = beta * np.asarray(win_de)
+                win_ts_arr = np.asarray(win_ts)
 
-            # convert last window's work and timestep values to np arrays
-            win_de_arr = beta * np.asarray(win_de)
-            win_ts_arr = np.asarray(win_ts)
+                # extract lambda values for finished window
+                # lambda1 = sampling lambda (row), lambda2 = evaluated lambda (col)
+                lambda1 = "{0:.2f}".format(float(l[7]))
+                lambda2 = "{0:.2f}".format(float(l[8]))
 
-            # extract lambda values for finished window
-            # lambda1 = sampling lambda (row), lambda2 = evaluated lambda (col)
-            lambda1 = "{0:.2f}".format(float(l[7]))
-            lambda2 = "{0:.2f}".format(float(l[8]))
+                # create dataframe of times and work values
+                # this window's data goes in row LAMBDA1 and column LAMBDA2
+                tempDF = pd.DataFrame({
+                    'time': win_ts_arr,
+                    'fep-lambda': np.full(len(win_de_arr),lambda1),
+                    lambda1: 0,
+                    lambda2: win_de_arr})
 
-            # create dataframe of timestep and work values
-            # this window's data goes in row LAMBDA1 and column LAMBDA2
-            tempDF = pd.DataFrame({
-                'timestep': win_ts_arr,
-                'fep-lambda': np.full(len(win_de_arr),lambda1),
-                lambda1: 0,
-                lambda2: win_de_arr})
+                # join the new window's df to existing df
+                u_nk = pd.concat([u_nk, tempDF], sort=True)
+                #u_nk.fillna(0, inplace=True)
 
-            # join the new window's df to existing df
-            u_nk = pd.concat([u_nk, tempDF])
-            u_nk.fillna(0, inplace=True)
+                # reset values for next window of fepout file
+                win_de = []
+                win_ts = []
+                parsing = False
 
-            # reset values for next window of fepout file
-            win_de = []
-            win_ts = []
-            parsing = False
+            # append work value from 'dE' column of fepout file
+            if parsing:
+                win_de.append(float(l[6]))
+                win_ts.append(float(l[1]))
 
-        # append work value from 'dE' column of fepout file
-        if parsing:
-            win_de.append(float(l[6]))
-            win_ts.append(float(l[1]))
-
-        # turn parsing on after line 'STARTING COLLECTION OF ENSEMBLE AVERAGE'
-        if '#STARTING' in l:
-            parsing = True
+            # turn parsing on after line 'STARTING COLLECTION OF ENSEMBLE AVERAGE'
+            if '#STARTING' in l:
+                parsing = True
 
     # create last dataframe for fep-lambda at last LAMBDA2
     tempDF = pd.DataFrame({
-        'timestep':win_ts_arr,
+        'time': win_ts_arr,
         'fep-lambda': lambda2})
-    u_nk = pd.concat([u_nk, tempDF])
-    u_nk.fillna(0, inplace=True)
 
-    u_nk.set_index(['timestep','fep-lambda'], inplace=True)
+    u_nk = pd.concat([u_nk, tempDF], sort=True)
+    #u_nk.fillna(0, inplace=True)
+
+    u_nk.set_index(['time','fep-lambda'], inplace=True)
 
     return u_nk
-

--- a/src/alchemlyb/parsing/namd.py
+++ b/src/alchemlyb/parsing/namd.py
@@ -5,13 +5,22 @@ import pandas as pd
 import numpy as np
 from .util import anyopen
 
-def extract_u_nk(fep_file):
+# TODO: perhaps move constants elsewhere?
+# these are the units we need for dealing with NAMD, which uses
+# kcal/mol for energies  http://www.ks.uiuc.edu/Research/namd/2.13/ug/node12.html#SECTION00062200000000000000
+# (kB in kcal/molK)
+k_b = 1.9872041e-3
+
+
+def extract_u_nk(fep_file, T):
     """Return reduced potentials `u_nk` from NAMD fepout file.
 
     Parameters
     ----------
     fep_file : str
         Path to fepout file to extract data from.
+    T : float
+        Temperature in Kelvin at which the simulation was sampled.
 
     Returns
     -------
@@ -19,6 +28,8 @@ def extract_u_nk(fep_file):
         Potential energy for each alchemical state (k) for each frame (n).
 
     """
+    beta = 1/(k_b * T)
+
     # lists to get timesteps and work values of each window
     win_ts = []
     win_de = []
@@ -40,7 +51,7 @@ def extract_u_nk(fep_file):
         if '#Free' in l:
 
             # convert last window's work and timestep values to np arrays
-            win_de_arr = np.asarray(win_de)
+            win_de_arr = beta * np.asarray(win_de)
             win_ts_arr = np.asarray(win_ts)
 
             # extract lambda values for finished window
@@ -51,10 +62,10 @@ def extract_u_nk(fep_file):
             # create dataframe of timestep and work values
             # this window's data goes in row LAMBDA1 and column LAMBDA2
             tempDF = pd.DataFrame({
-                'timestep':win_ts_arr,
+                'timestep': win_ts_arr,
                 'fep-lambda': np.full(len(win_de_arr),lambda1),
-                lambda1:0,
-                lambda2:win_de_arr})
+                lambda1: 0,
+                lambda2: win_de_arr})
 
             # join the new window's df to existing df
             u_nk = pd.concat([u_nk, tempDF])

--- a/src/alchemlyb/parsing/namd.py
+++ b/src/alchemlyb/parsing/namd.py
@@ -67,7 +67,6 @@ def extract_u_nk(fep_file, T):
 
                 # join the new window's df to existing df
                 u_nk = pd.concat([u_nk, tempDF], sort=True)
-                #u_nk.fillna(0, inplace=True)
 
                 # reset values for next window of fepout file
                 win_de = []
@@ -89,7 +88,6 @@ def extract_u_nk(fep_file, T):
         'fep-lambda': lambda2})
 
     u_nk = pd.concat([u_nk, tempDF], sort=True)
-    #u_nk.fillna(0, inplace=True)
 
     u_nk.set_index(['time','fep-lambda'], inplace=True)
 

--- a/src/alchemlyb/tests/parsing/test_namd.py
+++ b/src/alchemlyb/tests/parsing/test_namd.py
@@ -18,7 +18,7 @@ def test_u_nk(dataset, direction, shape):
     """Test that u_nk has the correct form when extracted from files.
     """
     for filename in dataset['data'][direction]:
-        u_nk = extract_u_nk(filename)
+        u_nk = extract_u_nk(filename, T=300)
 
         assert u_nk.index.names == ['timestep', 'fep-lambda']
         assert u_nk.shape == shape

--- a/src/alchemlyb/tests/parsing/test_namd.py
+++ b/src/alchemlyb/tests/parsing/test_namd.py
@@ -1,46 +1,24 @@
 """NAMD parser tests.
 
 """
+import pytest
 
 from alchemlyb.parsing.namd import extract_u_nk
 from alchemtest.namd import load_tyr2ala
-from numpy.testing import assert_almost_equal
 
+@pytest.fixture(scope="module")
+def dataset():
+    return load_tyr2ala()
 
-def test_u_nk():
+@pytest.mark.parametrize("direction,shape",
+                         [('forward', (21021, 21)),
+                          ('backward', (21021, 21)),
+                          ])
+def test_u_nk(dataset, direction, shape):
     """Test that u_nk has the correct form when extracted from files.
     """
-    dataset = load_tyr2ala()
+    for filename in dataset['data'][direction]:
+        u_nk = extract_u_nk(filename)
 
-    for direction in dataset['data']:
-        for filename in dataset['data'][direction]:
-            u_nk = extract_u_nk(filename)
-
-            assert u_nk.index.names == ['timestep', 'fep-lambda']
-            if direction == 'forward':
-                assert u_nk.shape == (21021, 21)
-            elif direction == 'backward':
-                assert u_nk.shape == (21021, 21)
-
-def test_bar_namd():
-    """Test BAR calculation on NAMD data.
-    """
-    from alchemlyb.estimators import BAR
-    import numpy as np
-
-    # load data
-    dataset = load_tyr2ala()
-    u_nk1 = extract_u_nk(dataset['data']['forward'][0])
-    u_nk2 = extract_u_nk(dataset['data']['backward'][0])
-
-    # combine dataframes of fwd and rev directions
-    u_nk1.replace(0, np.nan, inplace=True)
-    u_nk1[u_nk1.isnull()] = u_nk2
-    u_nk1.replace(np.nan, 0, inplace=True)
-    u_nk = u_nk1.sort_index(level=u_nk1.index.names[1:])
-
-    # after loading BOTH fwd and rev data, do BAR calculation
-    bar = BAR()
-    bar.fit(u_nk)
-    dg = (bar.delta_f_.iloc[0].iloc[-1])
-    assert_almost_equal(dg, 6.03126982925, decimal=4)
+        assert u_nk.index.names == ['timestep', 'fep-lambda']
+        assert u_nk.shape == shape

--- a/src/alchemlyb/tests/parsing/test_namd.py
+++ b/src/alchemlyb/tests/parsing/test_namd.py
@@ -20,5 +20,5 @@ def test_u_nk(dataset, direction, shape):
     for filename in dataset['data'][direction]:
         u_nk = extract_u_nk(filename, T=300)
 
-        assert u_nk.index.names == ['timestep', 'fep-lambda']
+        assert u_nk.index.names == ['time', 'fep-lambda']
         assert u_nk.shape == shape

--- a/src/alchemlyb/tests/test_fep_estimators.py
+++ b/src/alchemlyb/tests/test_fep_estimators.py
@@ -119,17 +119,22 @@ class TestMBAR(FEPestimatorMixin):
     """
     cls = MBAR
 
-    @pytest.mark.parametrize('X_delta_f', [
-        (gmx_benzene_coul_u_nk(), 3.041, 0.02088),
-        (gmx_benzene_vdw_u_nk(), -3.007, 0.04519),
-        (gmx_expanded_ensemble_case_1(), 75.923, 0.14124),
-        (gmx_expanded_ensemble_case_2(), 75.915, 0.14372),
-        (gmx_expanded_ensemble_case_3(), 76.173, 0.11345),
-        (gmx_water_particle_with_total_energy(), -11.680, 0.083655),
-        (gmx_water_particle_with_potential_energy(), -11.675, 0.083589),
-        (gmx_water_particle_without_energy(), -11.654, 0.083415),
-        (amber_bace_example_complex_vdw(), 2.40200, 0.0618453),
-    ])
+
+    @pytest.fixture(scope="class",
+                    params=[(gmx_benzene_coul_u_nk, 3.041, 0.02088),
+                            (gmx_benzene_vdw_u_nk, -3.007, 0.04519),
+                            (gmx_expanded_ensemble_case_1, 75.923, 0.14124),
+                            (gmx_expanded_ensemble_case_2, 75.915, 0.14372),
+                            (gmx_expanded_ensemble_case_3, 76.173, 0.11345),
+                            (gmx_water_particle_with_total_energy, -11.680, 0.083655),
+                            (gmx_water_particle_with_potential_energy, -11.675, 0.083589),
+                            (gmx_water_particle_without_energy, -11.654, 0.083415),
+                            (amber_bace_example_complex_vdw, 2.40200, 0.0618453),
+                    ])
+    def X_delta_f(self, request):
+        get_unk, E, dE = request.param
+        return get_unk(), E, dE
+
     def test_mbar(self, X_delta_f):
         self.compare_delta_f(X_delta_f)
 
@@ -142,18 +147,22 @@ class TestBAR(FEPestimatorMixin):
     T = 300
     kT_NAMD = namd.k_b * T
 
-    @pytest.mark.parametrize('X_delta_f', [
-        (gmx_benzene_coul_u_nk(), 3.044, 0.01640),
-        (gmx_benzene_vdw_u_nk(), -3.033, 0.03438),
-        (gmx_expanded_ensemble_case_1(), 75.993, 0.11056),
-        (gmx_expanded_ensemble_case_2(), 76.009, 0.11220),
-        (gmx_expanded_ensemble_case_3(), 76.219, 0.08886),
-        (gmx_water_particle_with_total_energy(), -11.675, 0.065055),
-        (gmx_water_particle_with_potential_energy(), -11.724, 0.064964),
-        (gmx_water_particle_without_energy(), -11.660, 0.064914),
-        (amber_bace_example_complex_vdw(), 2.37846, 0.050899),
-        (namd_tyr2ala(), 6.031269829/kT_NAMD, 0.069813058/kT_NAMD),
-    ])
+    @pytest.fixture(scope="class",
+                    params = [(gmx_benzene_coul_u_nk, 3.044, 0.01640),
+                              (gmx_benzene_vdw_u_nk, -3.033, 0.03438),
+                              (gmx_expanded_ensemble_case_1, 75.993, 0.11056),
+                              (gmx_expanded_ensemble_case_2, 76.009, 0.11220),
+                              (gmx_expanded_ensemble_case_3, 76.219, 0.08886),
+                              (gmx_water_particle_with_total_energy, -11.675, 0.065055),
+                              (gmx_water_particle_with_potential_energy, -11.724, 0.064964),
+                              (gmx_water_particle_without_energy, -11.660, 0.064914),
+                              (amber_bace_example_complex_vdw, 2.37846, 0.050899),
+                              (namd_tyr2ala, 6.031269829/kT_NAMD, 0.069813058/kT_NAMD),
+                    ])
+    def X_delta_f(self, request):
+        get_unk, E, dE = request.param
+        return get_unk(), E, dE
+
     def test_bar(self, X_delta_f):
         self.compare_delta_f(X_delta_f)
 

--- a/src/alchemlyb/tests/test_fep_estimators.py
+++ b/src/alchemlyb/tests/test_fep_estimators.py
@@ -3,13 +3,15 @@
 """
 import pytest
 
+import numpy as np
 import pandas as pd
 
-from alchemlyb.parsing import gmx, amber
+from alchemlyb.parsing import gmx, amber, namd
 from alchemlyb.estimators import MBAR
 from alchemlyb.estimators import BAR
 import alchemtest.gmx
 import alchemtest.amber
+import alchemtest.namd
 
 def gmx_benzene_coul_u_nk():
     dataset = alchemtest.gmx.load_benzene()
@@ -82,6 +84,19 @@ def amber_bace_example_complex_vdw():
                       for filename in dataset['data']['complex']['vdw']])
     return u_nk
 
+def namd_tyr2ala():
+    dataset = alchemtest.namd.load_tyr2ala()
+    u_nk1 = namd.extract_u_nk(dataset['data']['forward'][0])
+    u_nk2 = namd.extract_u_nk(dataset['data']['backward'][0])
+
+    # combine dataframes of fwd and rev directions
+    u_nk1.replace(0, np.nan, inplace=True)
+    u_nk1[u_nk1.isnull()] = u_nk2
+    u_nk1.replace(np.nan, 0, inplace=True)
+    u_nk = u_nk1.sort_index(level=u_nk1.index.names[1:])
+
+    return u_nk
+
 class FEPestimatorMixin:
     """Mixin for all FEP Estimator test classes.
 
@@ -134,6 +149,7 @@ class TestBAR(FEPestimatorMixin):
         (gmx_water_particle_with_potential_energy(), -11.724, 0.064964),
         (gmx_water_particle_without_energy(), -11.660, 0.064914),
         (amber_bace_example_complex_vdw(), 2.37846, 0.050899),
+        (namd_tyr2ala(), 6.031269829, 0.069813058),
     ])
     def test_bar(self, X_delta_f):
         self.compare_delta_f(X_delta_f)

--- a/src/alchemlyb/tests/test_fep_estimators.py
+++ b/src/alchemlyb/tests/test_fep_estimators.py
@@ -86,8 +86,8 @@ def amber_bace_example_complex_vdw():
 
 def namd_tyr2ala():
     dataset = alchemtest.namd.load_tyr2ala()
-    u_nk1 = namd.extract_u_nk(dataset['data']['forward'][0])
-    u_nk2 = namd.extract_u_nk(dataset['data']['backward'][0])
+    u_nk1 = namd.extract_u_nk(dataset['data']['forward'][0], T=300)
+    u_nk2 = namd.extract_u_nk(dataset['data']['backward'][0], T=300)
 
     # combine dataframes of fwd and rev directions
     u_nk1.replace(0, np.nan, inplace=True)
@@ -139,6 +139,9 @@ class TestBAR(FEPestimatorMixin):
     """
     cls = BAR
 
+    T = 300
+    kT_NAMD = namd.k_b * T
+
     @pytest.mark.parametrize('X_delta_f', [
         (gmx_benzene_coul_u_nk(), 3.044, 0.01640),
         (gmx_benzene_vdw_u_nk(), -3.033, 0.03438),
@@ -149,7 +152,7 @@ class TestBAR(FEPestimatorMixin):
         (gmx_water_particle_with_potential_energy(), -11.724, 0.064964),
         (gmx_water_particle_without_energy(), -11.660, 0.064914),
         (amber_bace_example_complex_vdw(), 2.37846, 0.050899),
-        (namd_tyr2ala(), 6.031269829, 0.069813058),
+        (namd_tyr2ala(), 6.031269829/kT_NAMD, 0.069813058/kT_NAMD),
     ])
     def test_bar(self, X_delta_f):
         self.compare_delta_f(X_delta_f)

--- a/src/alchemlyb/tests/test_fep_estimators.py
+++ b/src/alchemlyb/tests/test_fep_estimators.py
@@ -90,7 +90,7 @@ def namd_tyr2ala():
     u_nk2 = namd.extract_u_nk(dataset['data']['backward'][0], T=300)
 
     # combine dataframes of fwd and rev directions
-    u_nk1[u_nk1.isnan()] = u_nk2
+    u_nk1[u_nk1.isna()] = u_nk2
     u_nk = u_nk1.sort_index(level=u_nk1.index.names[1:])
 
     return u_nk
@@ -117,7 +117,6 @@ class TestMBAR(FEPestimatorMixin):
     """
     cls = MBAR
 
-
     @pytest.fixture(scope="class",
                     params=[(gmx_benzene_coul_u_nk, 3.041, 0.02088),
                             (gmx_benzene_vdw_u_nk, -3.007, 0.04519),
@@ -136,15 +135,12 @@ class TestMBAR(FEPestimatorMixin):
     def test_mbar(self, X_delta_f):
         self.compare_delta_f(X_delta_f)
 
+
 class TestBAR(FEPestimatorMixin):
     """Tests for BAR.
 
     """
     cls = BAR
-
-    # NAMD data were generated at T=300 K and the reference values
-    # were provided in kcal/mol so we convert to reduced units
-    kT_NAMD = 300 * namd.k_b
 
     @pytest.fixture(scope="class",
                     params = [(gmx_benzene_coul_u_nk, 3.044, 0.01640),
@@ -156,7 +152,7 @@ class TestBAR(FEPestimatorMixin):
                               (gmx_water_particle_with_potential_energy, -11.724, 0.064964),
                               (gmx_water_particle_without_energy, -11.660, 0.064914),
                               (amber_bace_example_complex_vdw, 2.37846, 0.050899),
-                              (namd_tyr2ala, 6.031269829/kT_NAMD, 0.069813058/kT_NAMD),
+                              (namd_tyr2ala, 11.0044, 0.10235),
                     ])
     def X_delta_f(self, request):
         get_unk, E, dE = request.param

--- a/src/alchemlyb/tests/test_fep_estimators.py
+++ b/src/alchemlyb/tests/test_fep_estimators.py
@@ -90,9 +90,7 @@ def namd_tyr2ala():
     u_nk2 = namd.extract_u_nk(dataset['data']['backward'][0], T=300)
 
     # combine dataframes of fwd and rev directions
-    u_nk1.replace(0, np.nan, inplace=True)
-    u_nk1[u_nk1.isnull()] = u_nk2
-    u_nk1.replace(np.nan, 0, inplace=True)
+    u_nk1[u_nk1.isnan()] = u_nk2
     u_nk = u_nk1.sort_index(level=u_nk1.index.names[1:])
 
     return u_nk
@@ -144,8 +142,9 @@ class TestBAR(FEPestimatorMixin):
     """
     cls = BAR
 
-    T = 300
-    kT_NAMD = namd.k_b * T
+    # NAMD data were generated at T=300 K and the reference values
+    # were provided in kcal/mol so we convert to reduced units
+    kT_NAMD = 300 * namd.k_b
 
     @pytest.fixture(scope="class",
                     params = [(gmx_benzene_coul_u_nk, 3.044, 0.01640),

